### DISCUSSION TITLE
Position estimator - velocityZ integration

### DISCUSSION THREAD

**Discussion by fynv** [bitcraze/discussions#930] [Q&A] [answered]
line 120 position_estimator_altitude.c, positionUpdateVelocityInternal():

`state->velocityZ += deadband(accWZ, state->vAccDeadband) * dt * G;`

I don't understand why is "G" multiplied here. Is it a mistake?

Thanks.

**Comment by knmcguire** [comment, marked as answer]
I've just converted your issue to discussions as it is more of a support question than an issue.

G stands for gravitational acceleration, so in this context in terms of position estimation, it does not seem weird to have a gravitational factor here. Why would you think it is a mistake?

Just for reference: this is the location of the code: 

https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/estimator/position_estimator_altitude.c#L120

**Reply by fynv** [reply]
"accWZ" is the estimated acceleration along Z axis. G is also an acceleration (in m/s^2 unit).
So it seems to me that the code is multiplying 2 acceleration terms together here..?

**Reply by fynv** [reply]
Sorry.. I didn't notice that the gyro read acc is in Gs. So yes, it makes perfect sense here.
(see definition of "struct state_s" in stabilizer_types.h).
Btw, I'm reading the code because I'm trying to write a general quadcopter simulator and I'm not famliar with the behavior of the sensor hardware.

**Reply by fynv** [reply]
What about gx/gy/gz? In "sensfusion6.c" I see there's hint that they are in degree/s rather than radian/s.
However, in "Madgwick's IMU and AHRS algorithms" version of code, I'm not sure. It may work both ways, is that correct?

**Reply by knmcguire** [reply]
you mean this code? https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/sensfusion6.c#L88

Here the gx,gy,gz are gyroscope measurements.

You are perhaps looking for gravX, gravY and gravZ, which stands for the unit factor that indicates the direction of gravity (compared to the orientation of the crazyflie I guess?)

https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/sensfusion6.c#L57

**Reply by fynv** [reply]
I meant the gyroscope measurements. It is clear to me now they are in deg/sec units. 

A concern here is that we need to convert them to rad/sec in sensfusion6Update.
There are 2 versions of "sensfusion6UpdateQImpl", Madgwick's and Mahony's.
In Mahony's version, there's conversion code:

```
  gx = gx * M_PI_F / 180;
  gy = gy * M_PI_F / 180;
  gz = gz * M_PI_F / 180;
```
But there's no such code in Madgwick's version.
In my simulator, Madgwick's version actually fails in current form. It works if I add the above code.
As to the firmware, I'm not sure if I missed something here. Maybe Madgwick's path is not used.

**Reply by knmcguire** [reply]
Ah that is indeed a bit concerning. To be honest, we haven't tried Madwick for a long time probably and maybe the back conversion to degrees got refactored out in the meantime but we never noticed it. In the kconfig file, Mahony is the standard: https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/Kconfig#L103

This seems to be a bug then so I'll open an issue to investigate it. perhaps better to discuss there since this discussion is already considered 'solved' as originally it was about the velocityZ intergration, so it is perhaps a bit unhandy to start a new branch of support in here. If you have a different question about the estimator, best to start a new discussion.

**Reply by knmcguire** [reply]
I've made an issue here: https://github.com/bitcraze/crazyflie-firmware/issues/1319. We should at one point check it out, but since this is a nonblocking issue it might take a while since we are pretty busy at the office.

Out of curiosity, for which simulator are you trying this code out🙂 ?

**Reply by fynv** [reply]
I'm writing one from scratch in C++.
I wrote a rendering engine previously https://github.com/fynv/three.v8. The simulator is based on it for rendering. 
For quadcopter simulation, I'm doing it by tracking the physical state of the rigid-body (position, rotation, velocity, and angluar-velocity), which is not too difficult.

**Reply by knmcguire** [reply]
ah very cool! If you'd like, you could share your project on 'show&tell' here on discussions 😄
